x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/query/SparseTermsQueryBuilder.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.query;
+
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.Query;
+import org.elasticsearch.Version;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.query.AbstractQueryBuilder;
+import org.elasticsearch.index.query.MatchNoneQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryRewriteContext;
+import org.elasticsearch.index.query.QueryShardException;
+import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+
+import static org.elasticsearch.common.Strings.format;
+import static org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper.requireNonNull;
+
+public class SparseTermsQueryBuilder extends AbstractQueryBuilder<SparseTermsQueryBuilder> {
+    public static final String NAME = "sparse_terms";
+
+    private static final ParseField FIELD = new ParseField("field");
+    private static final ParseField SCORES = new ParseField("scores");
+    private static final ParseField TERMS = new ParseField("terms");
+
+    @SuppressWarnings("unchecked")
+    public static final ConstructingObjectParser<SparseTermsQueryBuilder, Void> PARSER = new ConstructingObjectParser<>(
+        "sparse_terms",
+        false,
+        a -> new SparseTermsQueryBuilder((String) a[0], (List<String>) a[1], (List<Float>) a[2])
+    );
+
+    static {
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), FIELD);
+        PARSER.declareStringArray(ConstructingObjectParser.constructorArg(), TERMS);
+        PARSER.declareFloatArray(ConstructingObjectParser.constructorArg(), SCORES);
+        AbstractQueryBuilder.declareStandardFields(PARSER);
+    }
+
+    private final List<String> terms;
+    private final List<Float> scores;
+    private final String fieldName;
+
+    public SparseTermsQueryBuilder(String fieldName, List<String> terms, List<Float> scores) {
```

Comment:
We could reuse the `rank_feature` query instead of adding a new query, e.g.

```
{
  "bool": {
    "should": [
      { "rank_feature": { "field": "my_rank_features_field.term1", "boost": score1 } },
      ...
      { "rank_feature": { "field": "my_rank_features_field.termn", "boost": scoren } }
    ]
  }
}
```

If concatenating the name of the field with the term is annoying for some reasons, we may also be able to implement `termQuery` on `RankFeaturesFieldType` and then we could just use regular `term` queries to do this, ie.

```
{
  "bool": {
    "should": [
      { "term": { "my_rank_features_field": { "value": "term1", "boost": score1 } } },
      ...
      { "term": { "my_rank_features_field": { "value": "termn", "boost": scoren } } }
    ]
  }
}
```

Reply:
🤔

Reply:
I don't think we can reuse `term` with a boost as `term` will still score as BM25, and storing feature values as frequency would break this. So, we would have to make sure the mapped field type has the appropriate similarity configured (if we add a new "linear_frequency" similarity).

As for reusing `rank_feature` query, we could, but the linearCombination of rank feature would need to be updated in Lucene to allow weights greater than 64.

Reply:
🤦 I re-read your comment. 

Yes, I think returning my "SparseTermQuery" or whatever via the `termQuery` builder in RankedFeaturesMapper is a good idea. We then don't need another query builder, and this makes things much cleaner. 

I will also see about contributing some of these ideas to Lucene as I am sure others will want to use FeatureFields in this extensible manner in the future.